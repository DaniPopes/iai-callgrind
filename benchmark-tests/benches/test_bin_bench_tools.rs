use iai_callgrind::{
    binary_benchmark_group, main, Arg, BinaryBenchmarkConfig, Run, Tool, ValgrindTool,
};

#[inline(never)]
pub fn run_before() {
    std::fs::write("before", b"before").unwrap();
}

#[inline(never)]
pub fn run_after() {
    let string = std::fs::read_to_string("before").unwrap();
    assert_eq!(string, "before");
}

binary_benchmark_group!(
    name = subprocess;
    before = run_before, bench = true;
    after = run_after;
    benchmark = |"benchmark-tests-subprocess", group: &mut BinaryBenchmarkGroup| {
        group
            .bench(Run::with_arg(
                Arg::new(
                    "trace_sort_4000_sum_2000",
                    [
                        env!("CARGO_BIN_EXE_benchmark-tests-sort"),
                        &4000.to_string(),
                        &2000.to_string()
                    ],
                ),
            )
            .tool_override(Tool::new(ValgrindTool::DHAT)
                .args(["trace-children=yes"])
                .outfile_modifier("%p")
            )
        )
    }
);

binary_benchmark_group!(
    name = sort;
    before = run_before;
    after = run_after, bench = true;
    benchmark = |"benchmark-tests-sort", group: &mut BinaryBenchmarkGroup| {
        group
            .bench(Run::with_arg(
                Arg::new(
                    "sort_10_sum_10",
                    [
                        &10.to_string(),
                        &10.to_string()
                    ],
                ),
            ))
            .bench(Run::with_arg(
                Arg::new(
                    "sort_4000_sum_2000",
                    [
                        &4000.to_string(),
                        &2000.to_string()
                    ],
                ),
            )
        )
    }
);

// The main macro which creates a benchmarking harness with all group names from the above
// `binary_benchmark_group!` macros.
//
// We configure the regression checks for all benchmark groups with a percentage limit of `1%` for
// `Ir` (total instructions executed) and `10%` for `EstimatedCycles`. We also want to see all
// performance regressions and set `fail_fast` to false explicitly (This wouldn't have been
// necessary because the default is `false`). The whole benchmark still fails in the end if a
// performance regression was detected.
main!(
    config = BinaryBenchmarkConfig::default()
        .tool(Tool::new(ValgrindTool::DHAT))
        .tool(Tool::new(ValgrindTool::Massif))
        .tool(Tool::new(ValgrindTool::BBV))
        .tool(Tool::new(ValgrindTool::Memcheck))
        .tool(Tool::new(ValgrindTool::DRD))
        .tool(Tool::new(ValgrindTool::Helgrind));
    binary_benchmark_groups = sort, subprocess,
);
