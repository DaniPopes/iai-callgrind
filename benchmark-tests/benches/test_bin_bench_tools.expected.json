{
  "runs": [
    [
      {
        "group": "sort",
        "function": "benchmark-tests-sort",
        "id": "sort_10_sum_10",
        "expected": {
          "files": [
            "callgrind.benchmark-tests-sort.sort_10_sum_10.log",
            "callgrind.benchmark-tests-sort.sort_10_sum_10.out",
            "dhat.benchmark-tests-sort.sort_10_sum_10.log",
            "dhat.benchmark-tests-sort.sort_10_sum_10.out",
            "drd.benchmark-tests-sort.sort_10_sum_10.log",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.log",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.out.bb",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.out.pc",
            "helgrind.benchmark-tests-sort.sort_10_sum_10.log",
            "massif.benchmark-tests-sort.sort_10_sum_10.log",
            "massif.benchmark-tests-sort.sort_10_sum_10.out",
            "memcheck.benchmark-tests-sort.sort_10_sum_10.log",
            "summary.json"
          ]
        }
      },
      {
        "group": "sort",
        "function": "benchmark-tests-sort",
        "id": "sort_4000_sum_2000",
        "expected": {
          "files": [
            "callgrind.benchmark-tests-sort.sort_4000_sum_2000.log",
            "callgrind.benchmark-tests-sort.sort_4000_sum_2000.out",
            "dhat.benchmark-tests-sort.sort_4000_sum_2000.log",
            "dhat.benchmark-tests-sort.sort_4000_sum_2000.out",
            "drd.benchmark-tests-sort.sort_4000_sum_2000.log",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.log",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.out.bb",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.out.pc",
            "helgrind.benchmark-tests-sort.sort_4000_sum_2000.log",
            "massif.benchmark-tests-sort.sort_4000_sum_2000.log",
            "massif.benchmark-tests-sort.sort_4000_sum_2000.out",
            "memcheck.benchmark-tests-sort.sort_4000_sum_2000.log",
            "summary.json"
          ]
        }
      },
      {
        "group": "sort",
        "function": "run_after",
        "id": "after",
        "expected": {
          "files": [
            "callgrind.run_after.after.log",
            "callgrind.run_after.after.out",
            "dhat.run_after.after.log",
            "dhat.run_after.after.out",
            "drd.run_after.after.log",
            "exp-bbv.run_after.after.log",
            "exp-bbv.run_after.after.out.bb",
            "exp-bbv.run_after.after.out.pc",
            "helgrind.run_after.after.log",
            "massif.run_after.after.log",
            "massif.run_after.after.out",
            "memcheck.run_after.after.log",
            "summary.json"
          ]
        }
      },
      {
        "group": "subprocess",
        "function": "benchmark-tests-subprocess",
        "id": "trace_sort_4000_sum_2000",
        "expected": {
          "files": [
            "callgrind.benchmark-tests-subprocess.trace_sort_4000_sum_2000.log",
            "callgrind.benchmark-tests-subprocess.trace_sort_4000_sum_2000.out",
            "summary.json"
          ],
          "globs": [
            {
              "pattern": "dhat.benchmark-tests-subprocess.trace_sort_4000_sum_2000.log.*",
              "count": 2
            },
            {
              "pattern": "dhat.benchmark-tests-subprocess.trace_sort_4000_sum_2000.out.*",
              "count": 2
            }
          ]
        }
      },
      {
        "group": "subprocess",
        "function": "run_before",
        "id": "before",
        "expected": {
          "files": [
            "callgrind.run_before.before.log",
            "callgrind.run_before.before.out",
            "dhat.run_before.before.log",
            "dhat.run_before.before.out",
            "drd.run_before.before.log",
            "exp-bbv.run_before.before.log",
            "exp-bbv.run_before.before.out.bb",
            "exp-bbv.run_before.before.out.pc",
            "helgrind.run_before.before.log",
            "massif.run_before.before.log",
            "massif.run_before.before.out",
            "memcheck.run_before.before.log",
            "summary.json"
          ]
        }
      }
    ],
    [
      {
        "group": "sort",
        "function": "benchmark-tests-sort",
        "id": "sort_10_sum_10",
        "expected": {
          "files": [
            "callgrind.benchmark-tests-sort.sort_10_sum_10.log",
            "callgrind.benchmark-tests-sort.sort_10_sum_10.log.old",
            "callgrind.benchmark-tests-sort.sort_10_sum_10.out",
            "callgrind.benchmark-tests-sort.sort_10_sum_10.out.old",
            "dhat.benchmark-tests-sort.sort_10_sum_10.log",
            "dhat.benchmark-tests-sort.sort_10_sum_10.log.old",
            "dhat.benchmark-tests-sort.sort_10_sum_10.out",
            "dhat.benchmark-tests-sort.sort_10_sum_10.out.old",
            "drd.benchmark-tests-sort.sort_10_sum_10.log",
            "drd.benchmark-tests-sort.sort_10_sum_10.log.old",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.log",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.log.old",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.out.bb",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.out.bb.old",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.out.pc",
            "exp-bbv.benchmark-tests-sort.sort_10_sum_10.out.pc.old",
            "helgrind.benchmark-tests-sort.sort_10_sum_10.log",
            "helgrind.benchmark-tests-sort.sort_10_sum_10.log.old",
            "massif.benchmark-tests-sort.sort_10_sum_10.log",
            "massif.benchmark-tests-sort.sort_10_sum_10.log.old",
            "massif.benchmark-tests-sort.sort_10_sum_10.out",
            "massif.benchmark-tests-sort.sort_10_sum_10.out.old",
            "memcheck.benchmark-tests-sort.sort_10_sum_10.log",
            "memcheck.benchmark-tests-sort.sort_10_sum_10.log.old",
            "summary.json"
          ]
        }
      },
      {
        "group": "sort",
        "function": "benchmark-tests-sort",
        "id": "sort_4000_sum_2000",
        "expected": {
          "files": [
            "callgrind.benchmark-tests-sort.sort_4000_sum_2000.log",
            "callgrind.benchmark-tests-sort.sort_4000_sum_2000.log.old",
            "callgrind.benchmark-tests-sort.sort_4000_sum_2000.out",
            "callgrind.benchmark-tests-sort.sort_4000_sum_2000.out.old",
            "dhat.benchmark-tests-sort.sort_4000_sum_2000.log",
            "dhat.benchmark-tests-sort.sort_4000_sum_2000.log.old",
            "dhat.benchmark-tests-sort.sort_4000_sum_2000.out",
            "dhat.benchmark-tests-sort.sort_4000_sum_2000.out.old",
            "drd.benchmark-tests-sort.sort_4000_sum_2000.log",
            "drd.benchmark-tests-sort.sort_4000_sum_2000.log.old",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.log",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.log.old",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.out.bb",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.out.bb.old",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.out.pc",
            "exp-bbv.benchmark-tests-sort.sort_4000_sum_2000.out.pc.old",
            "helgrind.benchmark-tests-sort.sort_4000_sum_2000.log",
            "helgrind.benchmark-tests-sort.sort_4000_sum_2000.log.old",
            "massif.benchmark-tests-sort.sort_4000_sum_2000.log",
            "massif.benchmark-tests-sort.sort_4000_sum_2000.log.old",
            "massif.benchmark-tests-sort.sort_4000_sum_2000.out",
            "massif.benchmark-tests-sort.sort_4000_sum_2000.out.old",
            "memcheck.benchmark-tests-sort.sort_4000_sum_2000.log",
            "memcheck.benchmark-tests-sort.sort_4000_sum_2000.log.old",
            "summary.json"
          ]
        }
      },
      {
        "group": "sort",
        "function": "run_after",
        "id": "after",
        "expected": {
          "files": [
            "callgrind.run_after.after.log",
            "callgrind.run_after.after.log.old",
            "callgrind.run_after.after.out",
            "callgrind.run_after.after.out.old",
            "dhat.run_after.after.log",
            "dhat.run_after.after.log.old",
            "dhat.run_after.after.out",
            "dhat.run_after.after.out.old",
            "drd.run_after.after.log",
            "drd.run_after.after.log.old",
            "exp-bbv.run_after.after.log",
            "exp-bbv.run_after.after.log.old",
            "exp-bbv.run_after.after.out.bb",
            "exp-bbv.run_after.after.out.bb.old",
            "exp-bbv.run_after.after.out.pc",
            "exp-bbv.run_after.after.out.pc.old",
            "helgrind.run_after.after.log",
            "helgrind.run_after.after.log.old",
            "massif.run_after.after.log",
            "massif.run_after.after.log.old",
            "massif.run_after.after.out",
            "massif.run_after.after.out.old",
            "memcheck.run_after.after.log",
            "memcheck.run_after.after.log.old",
            "summary.json"
          ]
        }
      },
      {
        "group": "subprocess",
        "function": "benchmark-tests-subprocess",
        "id": "trace_sort_4000_sum_2000",
        "expected": {
          "files": [
            "callgrind.benchmark-tests-subprocess.trace_sort_4000_sum_2000.log",
            "callgrind.benchmark-tests-subprocess.trace_sort_4000_sum_2000.log.old",
            "callgrind.benchmark-tests-subprocess.trace_sort_4000_sum_2000.out",
            "callgrind.benchmark-tests-subprocess.trace_sort_4000_sum_2000.out.old",
            "summary.json"
          ],
          "globs": [
            {
              "pattern": "dhat.benchmark-tests-subprocess.trace_sort_4000_sum_2000.log.*",
              "count": 4
            },
            {
              "pattern": "dhat.benchmark-tests-subprocess.trace_sort_4000_sum_2000.out.*",
              "count": 4
            }
          ]
        }
      },
      {
        "group": "subprocess",
        "function": "run_before",
        "id": "before",
        "expected": {
          "files": [
            "callgrind.run_before.before.log",
            "callgrind.run_before.before.log.old",
            "callgrind.run_before.before.out",
            "callgrind.run_before.before.out.old",
            "dhat.run_before.before.log",
            "dhat.run_before.before.log.old",
            "dhat.run_before.before.out",
            "dhat.run_before.before.out.old",
            "drd.run_before.before.log",
            "drd.run_before.before.log.old",
            "exp-bbv.run_before.before.log",
            "exp-bbv.run_before.before.log.old",
            "exp-bbv.run_before.before.out.bb",
            "exp-bbv.run_before.before.out.bb.old",
            "exp-bbv.run_before.before.out.pc",
            "exp-bbv.run_before.before.out.pc.old",
            "helgrind.run_before.before.log",
            "helgrind.run_before.before.log.old",
            "massif.run_before.before.log",
            "massif.run_before.before.log.old",
            "massif.run_before.before.out",
            "massif.run_before.before.out.old",
            "memcheck.run_before.before.log",
            "memcheck.run_before.before.log.old",
            "summary.json"
          ]
        }
      }
    ]
  ]
}
